### PR TITLE
Directory Restructuring and SideInput Refresh Pipeline

### PR DESCRIPTION
**Pull request by prathapreddy123**
Its a massive PR with following changes:

1.  Restructured directories separating batch and streaming examples as well as enabling contributions to self contained pipelines with multiple modules/deployment scripts
2. Included CloudBuild.yaml files under each pipeline facilitating  to run tests for each pipeline in an isolated environment
3. Refactored slowlychanging-sideinput pattern ([PR](https://github.com/GoogleCloudPlatform/professional-services/pull/473) by expanding logic to multiple modules and written end to end pipeline tests

Note: Once PR is merged a reference to dataflow-python-examples/cloudbuild.yaml will be included in professional-services/cloudbuild.yaml

### CODE REVIEW

File: examples/dataflow-python-examples/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+   batch-examples/cookbook-examples or streaming-examples/cookbook-examples depending on the nature of pipeline.
+  
+*  Pipelines written either over multiple modules or has associated scripts (e.g: shell scripts/ terraform etc) must be
+   in its own directory as a sub directory under either batch-examples/ or streaming-examples/ depending on the nature of pipeline.
+   
+   Every self contained pipeline should follow below guidelines:
+    
+    1. Include requirements.txt specifying dependent python packages with respective versions 
+    
+    2. Include cloudbuild.yaml file containing 2 steps:
```
[the remark this answers is not in the document]
Comment (tswast) [reply examples/dataflow-python-examples/CONTRIBUTING.md:21]: I worry a bit about having two conflicting "contributing" steps, but makes sense to have some language / platform specific guides.

There's one potential issue with the way we are doing Cloud Build, which is that it seems like if there is any failure it'll cancel the whole pipeline, but I think since we actually trigger a Cloud Build job per cloudbuild.yaml it should be able to continue running the job even if we don't see the logs in the root job after one failure.

[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/CONTRIBUTING.md:21]: Initially I was leaning more towards having a single cloudbuild.yaml file for all examples but issue with that approach is all pipelines must be enforced to use same package versions

[the remark this answers is not in the document]
Comment (jaketf) [reply examples/dataflow-python-examples/CONTRIBUTING.md:21]: Not having the failure reason in the logs is far from an ideal contributing flow and will leave contributors asking reviewers to help them figure out why their check is failing.
In a future PR, could we write some helper that aggregates the logs from all children cloud build jobs run? this might be easier said than done.

File: examples/dataflow-python-examples/batch-examples/cookbook-examples/cloudbuild.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+steps:
```
Comment (jaketf) [inline examples/dataflow-python-examples/batch-examples/cookbook-examples/cloudbuild.yaml:1]: please add a Step 0: docker pull / cache to speed up these builds
https://cloud.google.com/cloud-build/docs/speeding-up-builds#using_a_cached_docker_image

note that cloud build zero-indexes it's steps so as-is these comments are off-by one.

File: examples/dataflow-python-examples/streaming-examples/cookbook-examples/README.md [hunk 3]
```diff
@@ -0,0 +1 @@
+# cookbook Streaming Pipelines
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/cookbook-examples/README.md:1]: Done

File: examples/dataflow-python-examples/.gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+/**/*.egg-info/
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/.gitignore:1]: Done

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
+This example demonstrates the concept of referencing multiple forms of side input data in GCS along with main input data 
+loaded from pubsub. Side input data can be refreshed periodically via PubSub notification.
+
+## UseCases
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/README.md:4]: Done

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/setup.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import setuptools
+
+setuptools.setup(
+    name='dataflow-slowlychanging-sideinput',
+    version='1.0',
+    author = 'Prathap Reddy',
+    author_email = 'parvathareddy@google.com',
+    install_requires=[],
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/setup.py:8]: Required to include only additional dependencies.  Beam is not required as dataflow runner will invoke appropriate sdk harness based on local user environment and installs additional dependencies listed here

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/sideinput_refresh/transforms.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+        Returns:
+         PCollection of type Json objects that represents data read from the files
+        """
+        # yapf: disable
+        return (pcol
+                | "Match files" >> beam.FlatMap(self._match_files).with_output_types(beamfs.FileMetadata)
+                | "Read Matches" >> beam.Map(
+                        lambda metadata: ReadableFile(metadata, self.compression)
+                    )  # Generates beam.io.fileio.ReadableFile
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/sideinput_refresh/transforms.py:87]: Done

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/sideinput_refresh/transforms.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+     Args:
+      event: Primary event data fetched from pubsub
+      name_field: Lookup key field name
+      value_field: Lookup value field name
+
+     Returns:
+         Look up information representing product name and lookup value
+    """
+
+    return (event[name_field], event[value_field])
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/sideinput_refresh/transforms.py:132]: Transform extracts values based on the specific schema. In this example assumption is sideinput data is conformed to schema specifications.  If we would like to handle a scenario for elements that doesn't meet schema spec then design should be changed to ignore such elements.

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""This script contains various tests to validate the pipeline end to end"""
+
+import unittest
+from sideinput_refresh import main,dofns,transforms,coders
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_pipeline.py:19]: Tests were formatted using yapf

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_pipeline.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+import functools
+from typing  import Tuple
+
+class TestCompletePipeline(unittest.TestCase):
+    def test_enrichment_allevents(self):
+         """Tests the pipeline by creating mock data and invoking respective transforms"""
+
+         with TestPipeline() as pipeline:
+            sideinput_filepath = (pipeline
+                                  | 'Read side Input file path' >> beam.Create(["gs://bucketname/input/sideinput/20200503".encode()])
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_pipeline.py:40]: Path doesn't matter as the data for tests are obtained from static_data.ReadStaticSideInput().   It's listed for illustration purpose how it appears in real world.

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+            """
+            if subscription == 'projects/test-project/subscriptions/sidesub':
+                return TestStream().add_elements([beam.window.TimestampedValue("path/to/sideinput/yyyymmddd".encode(), int(time.time()))])
+
+            return TestStream().add_elements(input_events)
+
+        @beam.ptransform_fn
+        def assertion_matcher(pcol, expected_values):
+            """Assertion matcher to match the pipeline output to expected output"""
+            assert_that(pcol,equal_to_per_window(expected_values),use_global_window=False,label='Assert events per window.')
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_pipeline.py:88]: Reformatted all test files using yapf

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_pipeline.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        with mock.patch('apache_beam.io.ReadFromPubSub', new=functools.partial(_mock_ReadFromPubSub, input_events = windowed_events)):
+            with mock.patch('sideinput_refresh.main.transforms.ReadSideInputData',new=static_data.ReadStaticSideInput):
+                main.run([
+                            '--project','test-project',
+                            '--runner','DirectRunner',
+                            '--sideinput-notification-subscription', 'projects/test-project/subscriptions/sidesub',
+                            '--maininput-subscription','projects/test-project/subscriptions/mainsub',
+                            '--sideinput-window-duration',"3600",
+                            '--maininput-window-duration',str(window_duration),
+                            '--demorun'
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_pipeline.py:104]: Done

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_transforms.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        with self.assertLogs(logger=None, level='INFO') as log:
+            utf8_jsoncoder.decode(message.encode("utf8"))
+            pattern  = "ERROR:root:Error (.+) in parsing value (.+)"
+            self.assertIsNotNone(re.match(pattern, log.output[0]),"Invalid utf8 input is not raising decode error")
+
+    def test_subscription_path_with_fullpath(self):
+        """Validate the format of subscription path when subscription name contains valid path"""
+        subsciption_path = main.get_subscription_path(
+            "project-id",
+            "projects/test-project/subscriptions/test-subscription")
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_transforms.py:45]: These 2 are independent tests representing two different ways subscription can be specified as input.  Each test can use its own values.  In this example I just used same subscription name as expected subscription name and hence appears to be repeated

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_transforms.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+        key_vaue_pair = transforms.kv_of(event, "product_name",
+                                                "discountpct")
+        self.assertTupleEqual(key_vaue_pair, ("product 2",0.35),
+                              "Mismatch in discountpct")
+
+        key_vaue_pair = transforms.kv_of(event, "product_name",
+                                                "category")
+        self.assertTupleEqual(key_vaue_pair, ("product 2", "electronics"),
+                              "Mismatch in category")
+
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/tests/test_transforms.py:82]: As explained earlier this transform extracts the values from an object where key's are expected to conform to a specific schema.

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/sideinput_refresh/transforms.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    event = json.loads(sales_event.decode())
+
+    event["bonuspoints"] = bonuspoints.get(event["productname"],
+                                           0) if bonuspoints else 0
+    event["discountpct"] = discountpct.get(event["productname"],
+                                           0.0) if discountpct else 0.0
+    event["category"] = category.get(event["productname"],
+                                     "Unknown") if category else "Unknown"
+
+    return (None, event)
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/sideinput_refresh/transforms.py:116]: Explained in the doc string

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/sideinput_refresh/transforms.py [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+
+     Returns:
+         Look up information representing product name and lookup value
+    """
+
+    return (event[name_field], event[value_field])
+
+
+@beam.typehints.with_input_types(Iterable[Dict[str, Any]])
+class EmitEvents(beam.DoFn):
```
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/sideinput_refresh/transforms.py:136]: I do see all types of naming conventions but however changed to indicated name

File: examples/dataflow-python-examples/batch-examples/cookbook-examples/cloudbuild.yaml (hunk 2 above)
Comment (prathapreddy123) [reply examples/dataflow-python-examples/batch-examples/cookbook-examples/cloudbuild.yaml:1]: --cached-from will be useful if we are persisting the image across builds. In the above example the **images** part pushes the image to GCR but in our case we are not persisting these images at the end of the build.  

Adjusted the step id indexes in comments to start with 0

Comment (jaketf) [reply examples/dataflow-python-examples/batch-examples/cookbook-examples/cloudbuild.yaml:1]: Is there a reason to not persist images across builds? this could speed up later runs.
For example:
PR 1 new example  -> builds test image for that example and pushes to GCR
PR 2 edit anything about the example without touching dependencies -> benefits from image being in GCR already

File: examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/setup.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaketf) [reply examples/dataflow-python-examples/streaming-examples/slowlychanging-sideinput/setup.py:8]: Interesting, thanks!

**Comment by jaketf** [comment]
@prathapreddy123  should we close #473 ?

File: examples/dataflow-python-examples/CONTRIBUTING.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (prathapreddy123) [reply examples/dataflow-python-examples/CONTRIBUTING.md]: Sure. We can explore that option..

File: examples/dataflow-python-examples/batch-examples/cookbook-examples/cloudbuild.yaml (hunk 2 above)
Comment (prathapreddy123) [reply examples/dataflow-python-examples/batch-examples/cookbook-examples/cloudbuild.yaml:1]: We can persist images as long as everybody is in agreement and do not have any concerns on the number and size of images.

File: examples/cryptorealtime/src/main/java/run/RunThePipeline.java [hunk 17]
```diff
@@ -135,7 +135,7 @@ public static void main(String[] args) {
         /* how to add additional custom pairs?
         bitFinexPair.add(new CurrencyPair("XTZ", "BTC"));*/
         allExchanges.add(new ExchangeConfiguration(BitfinexStreamingExchange.class.getName(),"bitfinex", bitFinexPair));
-
+	/*
```
Comment (jaketf) [inline examples/cryptorealtime/src/main/java/run/RunThePipeline.java:138]: @prathapreddy123 can you revert this unrelated change?
